docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,272 @@
+
+Prometheus also implements a [clean_tombstones](https://prometheus.io/docs/prometheus/latest/querying/api/#clean-tombstones) API which is not included in this proposal. The tombstones will be deleted automatically once the permanent deletion has taken place which is described in the section below. By default, this should take approximately 24 hours. 
+
+### Deletion Lifecycle
+
+The deletion request lifecycle can follow the same states as currently implemented for chunk storage:
+
+1. StatusReceived - No actions are done on request yet, just doing query time filtering
+2. StatusBuildingPlan - Request picked up for processing and building plans for it, still doing query time filtering
+3. StatusDeleting - Plans built already, running delete operations and still doing query time filtering
+4. StatusProcessed - All requested data deleted, not considering this for query time filtering
+
+(copied from the chunk storage series deletion proposal)
+
+With the current chunk store implementation, the amount of time for the request to move from StatusReceived to StatusBuildingPlan is dependent on a config option:  `-purger.delete-request-cancel-period`. The purpose of this is to allow the user some time to cancel the deletion request if it was made by mistake. 
+
+
+#### Current entities in the Purger that can be reused
+
+The following entities already exist for the chunk store and can be adapted to work for block storage:
+
+- `DeleteStore` - Handles storing and fetching delete requests, delete plans, cache gen numbers
+- `TombstonesLoader` - Helps loading delete requests and cache gen. Also keeps checking for updates.
+- `CacheGenMiddleware` - Adds generation number as a prefix to cache keys before doing Store/Fetch operations.
+
+(copied from the chunk storage series deletion proposal)
+
+
+
+### Filtering data during queries while not yet deleted:
+
+This will be done during the `DeleteRequest = StatusReceived` and `DeleteRequest = StatusBuildingPlan` parts of the deletion lifecycle.
+
+Once a deletion request is received, a tombstone entry will be created. The object store such as S3, GCS, Azure storage, can be used to store all the deletion requests. See the section below for more detail on how the tombstones will be stored. Using the tombstones, the querier will be able to filter the to-be-deleted data initially. In addition, the existing cache will be invalidated using cache generation numbers, which are described in the later sections. 
+
+From the purger service, the existing TombstonesLoader will periodically check for new tombstones in the object-store and load the new requests for deletion periodically using the modified DeleteStore. They will be loaded in memory periodically instead of retrieving all the tombstones when performing queries. Currently, with chunk storage, the TombstonesLoader is configured to check for updates to the DeleteStore every 5 minutes. 
```

Comment:
Are you suggesting that the purger updates the bucket index to append new tombstones? Wouldn't that lead to a possible race condition when a compactor for a specific tenant and the purger simultaneously write to a bucket index?

Would it make sense for the purger to create immutable tombstones separately and have the compactor append tombstones for a tenant to the bucket index whenever it updates it?

Reply:
I think it makes sense to do it how you suggested. 

One concern I have though is that it may take a while until the tombstones can be seen and used by the querier. Since the compactor will update the bucket index with the new tombstones every `-compactor.cleanup-interval` (default 15 min). Then the cached bucket index is refreshed in the querier every `-blocks-storage.bucket-store.sync-interval`(default 15 min). Potentially could take almost 30 min for queriers to start filtering deleted data. 

If this is okay, I'll go with the approach suggested. Otherwise, I might need to think of something else.

Reply:
> Are you suggesting that the purger updates the bucket index to append new tombstones?

No. The compactor is the only responsible to update the bucket index. As you said, the purger creates tombstones at `/<tenant id>/tombstones/<id>.json`, the the compactor list them and append them to the bucket index (as part of its bucket index update work).

Reply:
> One concern I have though is that it may take a while until the tombstones can be seen and used by the querier.

Yeah. I think we're missing a specification here. It's not clear how much we're willing to wait between when a tombstone is created and its data filtered out from queries.

Reply:
@pracucci - Are you concerned if we do a list objects by prefix call against `/<tenant_id>/tombstones/` every few minutes?

Reply:
Yes, I am. The reason is that I believe it will vanish the benefits of bucket-index, getting back the number of List operations to the bucket to the pre bucket-index era.